daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *metadata_filepath = get_metadata_filepath (self);
+  priv->metadata_key_file = g_key_file_new ();
+  GError *local_error = NULL;
+  gboolean load_succeeded =
+    g_key_file_load_from_file (priv->metadata_key_file, metadata_filepath,
+                               G_KEY_FILE_NONE, &local_error);
+  g_free (metadata_filepath);
+  if (!load_succeeded)
+    {
+      if (!g_error_matches (local_error, G_KEY_FILE_ERROR,
+                            G_KEY_FILE_ERROR_NOT_FOUND) &&
+          !g_error_matches (local_error, G_FILE_ERROR, G_FILE_ERROR_NOENT))
+        {
+          g_propagate_error (error, local_error);
+          return FALSE;
+        }
+
+      guint64 little_endian_max_size =
+        swap_bytes_64_if_big_endian (priv->max_size);
+      g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                             MAX_SIZE_KEY, little_endian_max_size);
+      return set_metadata (self, 0, 0, error);
+    }
+
+  guint64 little_endian_size =
+    g_key_file_get_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           SIZE_KEY, &local_error);
+  if (local_error != NULL)
+    goto handle_failed_read;
+
+  priv->size = swap_bytes_64_if_big_endian (little_endian_size);
+  if (priv->size > priv->max_size)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Logical size "
+                   "of circular file must be at most %" G_GUINT64_FORMAT ", "
+                   "but was %" G_GUINT64_FORMAT ".", priv->max_size,
+                   priv->size);
+      return FALSE;
+    }
+
+  goffset little_endian_key =
+    g_key_file_get_int64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                          HEAD_KEY, &local_error);
+  if (local_error != NULL)
+    goto handle_failed_read;
+
+  priv->head = swap_bytes_64_if_big_endian (little_endian_key);
+  if (priv->head < 0 || priv->head >= priv->max_size)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Pointer to "
+                   "head of circular file must lie in range [0, %"
+                   G_GUINT64_FORMAT "), but was %" G_GOFFSET_FORMAT ".",
+                   priv->max_size, priv->head);
+      return FALSE;
+    }
+
+  guint64 little_endian_prev_max_size =
+    g_key_file_get_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           MAX_SIZE_KEY, &local_error);
+  if (local_error != NULL)
+    goto handle_failed_read;
+
+  guint64 prev_max_size =
+    swap_bytes_64_if_big_endian (little_endian_prev_max_size);
+  return (prev_max_size == priv->max_size) ||
+         resize (self, prev_max_size, error);
+
+handle_failed_read:
+  g_propagate_error (error, local_error);
+  return FALSE;
+}
+
+static void
+emer_circular_file_initable_init (GInitableIface *iface)
+{
+  iface->init = emer_circular_file_may_fail_init;
```

Comment:
This being the function currently known as `emer_circular_file_may_fail_init` or `emer_circular_file_initable_init`?